go/proxyd/daisychain.go

```diff
@@ -0,0 +1,485 @@
+	"github.com/rs/cors"
+)
+
+var (
+	// Represents the chain id for the Optimism L2 mainnet
+	MainnetChainId = big.NewInt(10)
+	// Represents the chain id for the Optimism L2 public testnet
+	KovanChainId = big.NewInt(69)
+)
+
+// maxEpoch represents the maximum epoch
+// supported by the daisychain. If the RequestOptions
+// is sent with an epoch larger than this, it is an
+// error
+const maxEpoch = uint(6)
+
+// DaisyChainServer is the main struct representing
+// the daisy chain server. It has a reference to
+// each possible backend
+type DaisyChainServer struct {
+	rpcServer          *http.Server
+	maxBodySize        int64
+	authenticatedPaths map[string]string
+	epoch1             *Backend
+	epoch2             *Backend
+	epoch3             *Backend
+	epoch4             *Backend
+	epoch5             *Backend
+	epoch6             *Backend
+	bedrockCutoffBlock *big.Int
+	l2ChainID          *big.Int
+	upgrader           *websocket.Upgrader
+}
+
+// RequestOptions represents the optional
+// additional RPC parameter than can be sent
+// by a user to the daisychain server. The epoch
+// represents the backend that the request should be
+// sent to.
+type RequestOptions struct {
+	Epoch uint `json:"epoch"`
+}
+
+// NewDaisyChainServer will create a daisychain server
+func NewDaisyChainServer(
+	backends map[string]*Backend,
+	maxBodySize int64,
+	authenticatedPaths map[string]string,
+	l2ChainID *big.Int,
+	bedrockCutoffBlock *big.Int,
+) *DaisyChainServer {
+	srv := DaisyChainServer{
+		epoch1:             backends["epoch1"],
+		epoch2:             backends["epoch2"],
+		epoch3:             backends["epoch3"],
+		epoch4:             backends["epoch4"],
+		epoch5:             backends["epoch5"],
+		epoch6:             backends["epoch6"],
+		maxBodySize:        maxBodySize,
+		authenticatedPaths: authenticatedPaths,
+		bedrockCutoffBlock: bedrockCutoffBlock,
+		l2ChainID:          l2ChainID,
+		upgrader: &websocket.Upgrader{
+			HandshakeTimeout: 5 * time.Second,
+		},
+	}
+	return &srv
+}
+
+// StartDaisyChain will start a daisychain server
+// and is the entrypoint to the application
+func StartDaisyChain(config *Config) (func(), error) {
+	if err := config.ValidateDaisyChainBackends(); err != nil {
+		return func() {}, err
+	}
+
+	lim := NewLocalRateLimiter()
+	_, backendsByName, err := config.BuildBackends(lim)
+	if err != nil {
+		return func() {}, err
+	}
+
+	resolvedAuth, err := config.ResolveAuth()
+	if err != nil {
+		return func() {}, err
+	}
+
+	// parse the config
+	srv := NewDaisyChainServer(
+		backendsByName,
+		config.Server.MaxBodySizeBytes,
+		resolvedAuth,
+		config.Eth.L2ChainID,
+		config.Eth.BedrockCutoffBlock,
+	)
+
+	l2ChainID, err := srv.checkChainIDs()
+	if err != nil {
+		return func() {}, err
+	}
+	if srv.l2ChainID == nil {
+		srv.l2ChainID = l2ChainID
+	}
+
+	if srv.l2ChainID.Cmp(MainnetChainId) == 0 {
+		log.Info("running on mainnet")
+		// TODO(tynes): check and set the mainnet cutoff block
```

Comment:
I'd like to reduce the amount of config as possible to prevent any sort of misconfiguration issues that could cause requests going to the wrong backend. I can be sure that the argument can be passed but it should be optional